File: services/rendezvous-relay/internal/relay/authorizer.go
package relay

import (
	"context"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"errors"
	"strings"

	"github.com/anhydrous99/remote-davinci/protocol"
)

type AuthorizerEvent struct {
	Headers   map[string]string `json:"headers"`
	MethodARN string            `json:"methodArn"`
	RouteARN  string            `json:"routeArn"`
}

type AuthorizerResult struct {
	PrincipalID    string                   `json:"principalId"`
	PolicyDocument AuthorizerPolicyDocument `json:"policyDocument"`
	Context        map[string]any           `json:"context"`
}

type AuthorizerPolicyDocument struct {
	Version   string                      `json:"Version"`
	Statement []AuthorizerPolicyStatement `json:"Statement"`
}

type AuthorizerPolicyStatement struct {
	Action   string `json:"Action"`
	Effect   string `json:"Effect"`
	Resource string `json:"Resource"`
}

type endpointReader interface {
	GetEndpoint(context.Context, string) (*Endpoint, error)
}

var errUnauthorized = errors.New("Unauthorized")

func NewAuthorizer(store endpointReader) func(context.Context, AuthorizerEvent) (AuthorizerResult, error) {
	return func(ctx context.Context, event AuthorizerEvent) (AuthorizerResult, error) {
		header := ""
		for name, value := range event.Headers {
			if strings.EqualFold(name, "authorization") {
				header = value
				break
			}
		}
		authorization, err := protocol.ParseAuthorization(header)
		if err != nil {
			return AuthorizerResult{}, errUnauthorized
		}
		contextValues := map[string]any{"authMode": "pairing"}
		principalID := "pairing"
		if authorization.Scheme == "bearer" {
			endpoint, err := store.GetEndpoint(ctx, authorization.EndpointID)
			if err != nil {
				return AuthorizerResult{}, err
			}
			hash := credentialDigest(authorization.Secret)
			if endpoint == nil || endpoint.RevokedAt != 0 || !constantTimeEqual(endpoint.CredentialHash, hash) {
				return AuthorizerResult{}, errUnauthorized
			}
			principalID = authorization.EndpointID
			contextValues = map[string]any{
				"authMode": "endpoint", "endpointId": authorization.EndpointID, "credentialHash": hash,
			}
		}
		resource := event.MethodARN
		if resource == "" {
			resource = event.RouteARN
		}
		if resource == "" {
			return AuthorizerResult{}, errUnauthorized
		}
		return AuthorizerResult{
			PrincipalID: principalID,
			PolicyDocument: AuthorizerPolicyDocument{
				Version:   "2012-10-17",
				Statement: []AuthorizerPolicyStatement{{Action: "execute-api:Invoke", Effect: "Allow", Resource: resource}},
			},
			Context: contextValues,
		}, nil
	}
}

func credentialDigest(secret string) string {
	decoded, _ := base64.RawURLEncoding.DecodeString(secret)
	digest := sha256.Sum256(decoded)
	return base64.RawURLEncoding.EncodeToString(digest[:])
}

func constantTimeEqual(left, right string) bool {
	return len(left) == len(right) && subtle.ConstantTimeCompare([]byte(left), []byte(right)) == 1
}

File: services/rendezvous-relay/src/authorizer.ts
import { createHash, timingSafeEqual } from "node:crypto";
import { parseAuthorization } from "@remote-davinci/protocol";
import { DynamoStore, type Store } from "./store.js";

interface AuthorizerEvent {
  headers?: Record<string, string | undefined>;
  methodArn?: string;
  routeArn?: string;
}

interface AuthorizerResult {
  principalId: string;
  policyDocument: {
    Version: "2012-10-17";
    Statement: [{ Action: "execute-api:Invoke"; Effect: "Allow"; Resource: string }];
  };
  context:
    | { authMode: "pairing" }
    | { authMode: "endpoint"; endpointId: string; credentialHash: string };
}

const unauthorized = (): never => { throw new Error("Unauthorized"); };
const digest = (secret: string): string => createHash("sha256").update(Buffer.from(secret, "base64url")).digest("base64url");
const equal = (a: string, b: string): boolean => {
  const left = Buffer.from(a);
  const right = Buffer.from(b);
  return left.length === right.length && timingSafeEqual(left, right);
};

export function createAuthorizer(store: Pick<Store, "getEndpoint">) {
  return async (event: AuthorizerEvent): Promise<AuthorizerResult> => {
    const header = Object.entries(event.headers ?? {}).find(([name]) => name.toLowerCase() === "authorization")?.[1];
    let authorization;
    try {
      authorization = parseAuthorization(header);
    } catch {
      return unauthorized();
    }

    let endpointId = "";
    if (authorization.scheme === "bearer") {
      endpointId = authorization.endpointId;
      const endpoint = await store.getEndpoint(endpointId);
      if (!endpoint || endpoint.revokedAt || !equal(endpoint.credentialHash, digest(authorization.secret))) {
        return unauthorized();
      }
    }

    const resource = event.methodArn ?? event.routeArn;
    if (!resource) return unauthorized();
    const context = authorization.scheme === "bearer"
      ? { authMode: "endpoint" as const, endpointId, credentialHash: digest(authorization.secret) }
      : { authMode: "pairing" as const };
    return {
      principalId: endpointId || "pairing",
      policyDocument: {
        Version: "2012-10-17",
        Statement: [{ Action: "execute-api:Invoke", Effect: "Allow", Resource: resource }],
      },
      context,
    };
  };
}

const store = new DynamoStore(process.env.TABLE_NAME ?? "missing");
export const handler = createAuthorizer(store);
